java/code/src/com/redhat/rhn/frontend/servlets/AuthorizationFilter.java

```diff
@@ -0,0 +1,198 @@
+            }
+            else if (hreq.getServletPath().endsWith(".do") || hreq.getServletPath().endsWith(".jsp")) {
+                handleStrutsAccess(hreq, noAuthEndpoints);
+            }
+            else {
+                handleSparkAccess(hreq, noAuthEndpoints);
+            }
+            LOG.debug("Access granted for user '{}' to URI '{}' [{}]",
+                    new RequestContext(hreq).getCurrentUser(), hreq.getRequestURI(), hreq.getMethod());
+        }
+        catch (PermissionException e) {
+            // TODO: Handle PermissionExceptions properly depending on the "Content-Type" and "Accept" headers
+            // TODO: possibly pass it down another filter
+            // TODO: Review PageFilter
+            LOG.debug("Access restricted for user '{}' to URI '{}' [{}]",
+                    new RequestContext(hreq).getCurrentUser(), hreq.getRequestURI(), hreq.getMethod());
+            if (ConfigDefaults.get().isRbacEnabled()) {
+                hres.sendError(HttpServletResponse.SC_FORBIDDEN, e.getMessage());
+                return;
+            }
+        }
+
+        // Pass control on to the next filter
+        chain.doFilter(request, response);
+    }
+
+    private void handleAjaxAccess(HttpServletRequest hreq, Set<String> noAuthEndpoints) {
+        String path = hreq.getServletPath() + hreq.getPathInfo();
+        if (!authenticationService.requestURIRequiresAuthentication(hreq) || noAuthEndpoints.contains(path)) {
+            return;
+        }
+
+        User user = new RequestContext(hreq).getCurrentUser();
+        if (user == null) {
+            if (!noAuthEndpoints.contains(path)) {
+                throw new PermissionException("The " + hreq.getRequestURI() +
+                        " is not available for unauthenticated users.");
+            }
+        }
+        else {
+            if (!user.hasRole(RoleFactory.SAT_ADMIN)) {
+                Optional<WebEndpoint> endpointOpts = WebEndpointFactory.lookupByUserIdEndpointScope(user.getId(),
+                        path,
+                        hreq.getMethod(),
+                        WebEndpoint.Scope.W);
+                if (endpointOpts.isEmpty()) {
+                    throw new PermissionException("The " + hreq.getRequestURI() +
+                            " API is not available to user " + user.getLogin());
+                }
+            }
+        }
+    }
+
+    private void handleStrutsAccess(HttpServletRequest hreq, Set<String> noAuthEndpoints) {
+        // Check if the request URI needs authentication. If so, also check if the URI needs authorization as well.
+        if (!authenticationService.requestURIRequiresAuthentication(hreq) ||
+                noAuthEndpoints.contains(hreq.getServletPath())) {
+            return;
+        }
+
+        User user = new RequestContext(hreq).getCurrentUser();
+        if (user == null) {
+            if (!noAuthEndpoints.contains(hreq.getServletPath())) {
```

Comment:
Same comment from the `handleSparkAccess`, up to you if you want to keep this check.